pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1369,24 +1403,31 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
 			glog.Error(errMessage)
 			return ErrExistingPodsAntiAffinityRulesNotMatch, errors.New(errMessage)
 		}
-		if matchingTerms, err = c.getMatchingAntiAffinityTerms(pod, filteredPods); err != nil {
+		if matchingTerms, topologyValuesToMatchingPods, err = c.getMatchingAntiAffinityTerms(pod, filteredPods); err != nil {
 			errMessage := fmt.Sprintf("Failed to get all terms that pod %+v matches, err: %+v", podName(pod), err)
 			glog.Error(errMessage)
 			return ErrExistingPodsAntiAffinityRulesNotMatch, errors.New(errMessage)
 		}
 	}
-	for _, terms := range matchingTerms {
-		for i := range terms {
-			term := &terms[i]
-			if len(term.term.TopologyKey) == 0 {
-				errMessage := fmt.Sprintf("Empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
-				glog.Error(errMessage)
-				return ErrExistingPodsAntiAffinityRulesNotMatch, errors.New(errMessage)
-			}
-			if priorityutil.NodesHaveSameTopologyKey(node, term.node, term.term.TopologyKey) {
-				glog.V(10).Infof("Cannot schedule pod %+v onto node %v,because of PodAntiAffinityTerm %v",
-					podName(pod), node.Name, term.term)
-				return ErrExistingPodsAntiAffinityRulesNotMatch, nil
+
+	// Iterate over topology values, to get matching pods and get their matching terms to check for same topolgy key
+	// currently ignored if predicateMetadata is not precomputed
```

Comment:
In the current logic, for each anti-affinity term, you store topology value and the associated set of pods in `metadata`. I think the logic in this function could be simplified if we store both topology key and value in `metadata`. If we store both key and value of the matched anti-affinity terms, we can check here to see if the node.Labels match any of those stored key/values stored. If any matches, the `node` is a match and fails the anti-affinity rules. We no longer need to iterate over `potentialPods` or check if `NodesHaveSameTopologyKey`.

If what I just wrote is correct, then we probably do not need to store `matchingAntiAffinityTerms` in metadata.
Let me know if I am missing anything.